EIPS/eip-3540.md

```diff
@@ -93,88 +90,94 @@ The list of section headers is terminated with the *section headers terminator b
+| kind_data              | 1 byte   | 0x04          | kind marker for data size section                                                                            |
+| data_size              | 2 bytes  | 0x0000-0xFFFF | 16-bit unsigned big-endian integer denoting the length of the data section content (*)                       |
+| terminator             | 1 byte   | 0x00          | marks the end of the header                                                                                  |
+
+(*) For not yet deployed containers this can be greater than the actual content length.
+
+#### Body
+
+| name              | length   | value         | description                                                                                |
+|-------------------|----------|---------------|--------------------------------------------------------------------------------------------|
+| types_section     | variable | n/a           | stores code section metadata                                                               |
+| inputs            | 1 byte   | 0x00-0x7F     | number of stack elements the code section consumes                                         |
+| outputs           | 1 byte   | 0x00-0x7F     | number of stack elements the code section returns                                          |
+| max_stack_height  | 2 bytes  | 0x0000-0x03FF | maximum number of elements ever placed onto the operand stack by the code section          |
+| code_section      | variable | n/a           | arbitrary bytecode                                                                         |
+| container_section | variable | n/a           | arbitrary EOF-formatted container                                                          |
+| data_section      | variable | n/a           | arbitrary sequence of bytes                                                                |
 
 See [EIP-4750](./eip-4750.md) for more information on the type section content.
 
 #### EOF version 1 validation rules
 
-1. In addition to general validation rules above, EOF version 1 bytecode conforms to the rules specified below:
-   - Exactly one type section header MUST be present immediately following the EOF version. Each code section MUST have a specified type signature in the type body.
-   - Exactly one code section header MUST be present immediately following the type section. A maximum of 1024 individual code sections are allowed.
-   - Exactly one data section header MUST be present immediately following the code section.
-2. Any version other than `0x01` is invalid.
+The following validity constraints are placed on the container format:
 
-(*Remark:* Contract creation code SHOULD set the section size of the data section so that the constructor arguments fit it.)
+- `version` must be `0x01`
+- `types_size` is divisible by `4`
+- the number of code sections must be equal to `types_size / 4`
+- the number of code sections must not exceed `1024`
```

Comment:
It's mentioned in the table, but maybe worth to mention here explicitly that `num_code_sections > 0`.